### PR TITLE
fix (jkube-kit/build) : ImageNameFormatter should respect `+` in image tag as per Semver specifications (#2301)

### LINKED ISSUE #2301: `k8s:build` fails to build image when image tag contains `+`
### Describe the bug

_originally posted by @cimnine in #1859_

If we configure Kubernetes Maven Plugin to use image name like this:
```xml
<jkube.generator.name>%a:4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455</jkube.generator.name>
```

`k8s:build` fails with the following error saying image name is invalid:

The given version is a valid SemVer version, however, it's not a valid image tag name. We should replace `+` character with some other character to handle this scenario.

### Eclipse JKube version

1.13.1

### Component

Kubernetes Maven Plugin

### Apache Maven version

other (please specify in additional context)

### Gradle version

None

### Steps to reproduce

Change image name generated by jkube to `%a:4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455` using property `jkube.generator.name`

### Expected behavior

`k8s:build` should handle `+` tag by replacing it with some other docker compatible token

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.25.3

### Environment

Linux

### Eclipse JKube Logs

```shell
[ERROR] k8s: Failed to execute the build [Given Docker name 'random-generator:4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455' is invalid:
   * tag part '4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455' doesn't match allowed pattern '^[\w][\w.-]{0,127}$'
See http://bit.ly/docker_image_fmt for more details]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  24.399 s
[INFO] Finished at: 2023-07-24T16:05:50+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.14-SNAPSHOT:build (default-cli) on project random-generator: Failed to execute the build: Given Docker name 'random-generator:4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455' is invalid:
[ERROR]    * tag part '4.3.3-SNAPSHOT+st1666023054591-b24-7ca68455' doesn't match allowed pattern '^[\w][\w.-]{0,127}$'
[ERROR] See http://bit.ly/docker_image_fmt for more details
```

### Sample Reproducer Project

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fixes #2301

_originally authored by @cimnine_ https://github.com/eclipse/jkube/pull/1859

I've just rebased the branch against latest master

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [ ] I have performed a self-review of my code
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java [hunk 1]
```diff
@@ -112,6 +115,95 @@ void tagWithNonSnapshotArtifact() {
...
+        project.setProperties(new Properties());
+        assertThat(formatter.format("%g/%a:%l"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:latest")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%v"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3-SNAPSHOT")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%t"))
+            .matches("^jkube/kubernetes-maven-plugin:snapshot-\\d{6}-\\d{6}-\\d{4}$")
+            .satisfies(ImageNameFormatterTest::validImageName);
```
Comment (manusa) [inline jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:132]: This should be a parameterized test

File: jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java [hunk 2]
```diff
@@ -112,6 +115,95 @@ void tagWithNonSnapshotArtifact() {
...
+        project.setProperties(properties);
+        assertThat(formatter.format("%g/%a:%l"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3_semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%v"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3_semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%t"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3_semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
```
Comment (manusa) [inline jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:168]: Should be parameterized test

File: jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java [hunk 3]
```diff
@@ -112,6 +115,95 @@ void tagWithNonSnapshotArtifact() {
...
+        project.setProperties(properties);
+        assertThat(formatter.format("%g/%a:%l"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3-semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%v"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3-semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%t"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3-semver.build_meta-data")
+            .satisfies(ImageNameFormatterTest::validImageName);
```
Comment (manusa) [inline jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:187]: should be parameterized test

File: jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java [hunk 4]
```diff
@@ -112,6 +115,95 @@ void tagWithNonSnapshotArtifact() {
...
+        project.setProperties(new Properties());
+        assertThat(formatter.format("%g/%a:%l"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%v"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3")
+            .satisfies(ImageNameFormatterTest::validImageName);
+        assertThat(formatter.format("%g/%a:%t"))
+            .isEqualTo("jkube/kubernetes-maven-plugin:1.2.3")
+            .satisfies(ImageNameFormatterTest::validImageName);
```
Comment (manusa) [inline jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:204]: should be parameterized test

File: jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java [hunk 5]
```diff
@@ -131,4 +223,8 @@ void format_whenPropertyInImageName_thenResolveProperty() {
         // Then
         assertThat(result).isEqualTo("registry.gitlab.com/myproject/myrepo/mycontainer:der12");
     }
+
+    private static void validImageName(String v) {
+        assertThatCode(() -> ImageName.validate(v)).doesNotThrowAnyException();
```
Comment (manusa) [inline jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:228]: Is this an assertion, an action, both? Can we write a specific test for this?
Can we move the assertion part to other scenarios?

Comment (rohanKanojia) [reply jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/helper/ImageNameFormatterTest.java:228]: this seems to be an assertion. `ImageName.validate` creates a new ImageName object which throws IllegalArgumentException if invalid name is provided

**Review by manusa — APPROVED** [review]

**Pull request merged**